### PR TITLE
Proof of concept task directory

### PR DESCRIPTION
**Pull request by eenblam**
Just experimenting with breaking out each task into a subdirectory.

Some things I'm doing that we don't have to stick with:

* Use `make`
* Use `./<taskname>/` format

We could re-org in different ways, for example:
```
tasks
├── 00_setup
├── 01_cases
├── 02_allegations
├── 03_pages
├── 04_participants
└── 05_docket
```

and have a top-level `Makefile` or script that walks through the task directories in numerical order, halting if any fails their internal process.

### CODE REVIEW

**Comment by neverett** [comment]
I like the idea of having a series of task-specific directories and a top-level Makefile. How would shared files and/or directories work in that case? Thinking of `db_config.py`, `connection.py`, and the like.

**Comment by eenblam** [comment]
Okay, this is now a decent proof of concept for how we could break out distinct tasks. I've added a very simple Makefile that just provides commands for specific, well, tasks to complete the abstract allegations "task." To generalize this:

* `make setup` should handle initial steps like creating the database table
* `make pre` should run tests that ought to pass prior to running `make main`
* `make main` should do the actual work of the migration
* `make post` should test that the `main` sub-task completed successfully. It currently doesn't revert, but it *would* fail a pipeline from proceeding to the next task in the pipeline.

If `make post` fails, `make clean` can be manually run to undo any changes the task may have made to the database. (This is similar to something like the `down` method in a Rails migration.) I think this will often just be "drop the new table we're building," but we currently also need to do things like re-NULL the `<whatever>_parse_error` columns in the `cases` table.

Curious what folks think! I'm happy to experiment further.

**Comment by eenblam** [comment]
> How would shared files and/or directories work in that case? Thinking of db_config.py, connection.py, and the like.

@neverett I've been using a rather ugly hack for `db_config`:

```python3
import sys, os
sys.path.append(os.getcwd() + '/..')
import db_config
```

---

BUT I just pushed some code that fixes that! See changes from 2b3fdbc for specifics, but in short, I created a `Common` package for tasks to import from.
 
* Created `Common/`
* Exposed modules within `Common/` via its `__init__.py` to save us from writing nested imports
* Added `allegations/common.py` as a local proxy module for `Common`
* Simplified imports in task scripts: just do `from common import Connection, db_config` :)

---

So, if anyone is creating a task:
* Just copy `allegations/common.py` into the new task
* Import within the task as `from common import Connection, db_config`

That's it! And if you want to add a module to `Common/`, then it can be helpful to account for it in `Common/__init__.py`.

Current contents, for reference:

```
(venv) ben@tin-can:~/b/gh/neverett/nlrb_data$ cat Common/__init__.py 
from .connection import Connection
import db_config
```

So, that should sort us out for shared Python code. As for other shared resources... maybe just a `data/` directory, with subdirectories within it as needed?

**Review by neverett — COMMENTED** [review]
one other small comment: you may want to update `allegations.sql` to use `CREATE TABLE IF NOT EXISTS` instead of `DROP TABLE IF EXISTS; CREATE TABLE...` -- the latter is what i originally had in the `cases` DDL, but refactored it so table creation isn't so unexpectedly catastrophic.

File: allegations/post.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+            print('Expected cases table to be populated, found 0 records')
+    except pymysql.err.ProgrammingError as e:
+        print(f'Could not count cases: {e}')
+
+    try:
+        c.execute("SELECT COUNT(*) c FROM allegations WHERE parse_error = 1;")
+        count = c.fetchone()['c']
+        if count != 0:
+            error = True
+            print(f"Expected {expected} allegations, found {count}")
```
[the remark this answers is not in the document]
Comment (eenblam) [reply allegations/post.py]: Yep, that should be `f"Expected 0 allegations, found {count}"`. Fixed now. Good catch!

**Comment by eenblam** [comment]
> one other small comment: you may want to update `allegations.sql` to use `CREATE TABLE IF NOT EXISTS` instead of `DROP TABLE IF EXISTS; CREATE TABLE...` -- the latter is what i originally had in the `cases` DDL, but refactored it so table creation isn't so unexpectedly catastrophic.

Good call! Fixed in [c272088](https://github.com/neverett/nlrb_data/pull/7/commits/c272088ae0cb2750329e2cb989b702760dc3d0b5)

**Comment by eenblam** [comment]
State of affairs at present:

* Common stuff lives in `Common/`
* Broke things out into `tasks/XX_name/` format
* Added path and SQL utilities to `Common/`
* Refactored `Common/connection.py` to just use the default cursor. `Connection().dict_cursor()` now provides a DictCursor instead.

Next steps, presumably post-merge:

* Sync up with @neverett 's task flow and work on new raw table ingest
    * Will also need to refactor allegations work to account for that
* Write a top-level task runner tool (maybe several things under `bin/`) (#9)
* Switch to PostgreSQL
* Additional parsers: participants (#4 ) and docket (#6 )
* Future:
    * Fuzzy search with PostgreSQL (#2 )
    * Consider supporting multiple DB backends, especial SQLite3 (#8)
    * Data analysis

**Review by neverett — APPROVED** [review]
LGTM, feel free to merge!

**Review by tgj505 — APPROVED** [review]

**Pull request merged**